pkg/runtime/grpc/oci.go

```diff
@@ -0,0 +1,290 @@
+				continue
+			}
+		}
+	})
+
+	targets, err := r.getTargets(gadgetCtx.Context(), r.ParamDescs().ToParams()) // gadgetCtx.RuntimeParams()) // TODO
+	if err != nil {
+		return fmt.Errorf("getting target nodes: %w", err)
+	}
+	_, err = r.runOCIGadgetOnTargets(gadgetCtx, paramMap, targets)
+	return err
+}
+
+func (r *Runtime) runOCIGadgetOnTargets(
+	gadgetCtx runtime.GadgetContext,
+	paramMap map[string]string,
+	targets []target,
+) (runtime.CombinedGadgetResult, error) {
+	results := make(runtime.CombinedGadgetResult)
+	var resultsLock sync.Mutex
+
+	wg := sync.WaitGroup{}
+	for _, t := range targets {
+		wg.Add(1)
+		go func(target target) {
+			gadgetCtx.Logger().Debugf("running oci gadget on node %q", target.node)
+			res, err := r.runOCIGadget(gadgetCtx, target, paramMap)
+			resultsLock.Lock()
+			results[target.node] = &runtime.GadgetResult{
+				Payload: res,
+				Error:   err,
+			}
+			resultsLock.Unlock()
+			wg.Done()
+		}(t)
+	}
+
+	wg.Wait()
+	return results, results.Err()
+}
+
+func (r *Runtime) runOCIGadget(gadgetCtx runtime.GadgetContext, target target, allParams map[string]string) ([]byte, error) {
+	// Notice that we cannot use gadgetCtx.Context() here, as that would - when cancelled by the user - also cancel the
+	// underlying gRPC connection. That would then lead to results not being received anymore (mostly for profile
+	// gadgets.)
+	connCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	timeout := time.Second * time.Duration(r.globalParams.Get(ParamConnectionTimeout).AsUint16())
+	dialCtx, cancelDial := context.WithTimeout(gadgetCtx.Context(), timeout)
+	defer cancelDial()
+
+	conn, err := r.dialContext(dialCtx, target, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("dialing target on node %q: %w", target.node, err)
+	}
+	defer conn.Close()
+	client := api.NewGadgetManagerClient(conn)
+
+	runRequest := &api.OCIGadgetRunRequest{
+		Url:      gadgetCtx.Args()[0],
+		Params:   allParams,
+		Args:     gadgetCtx.Args(),
+		Nodes:    nil,
+		LogLevel: uint32(gadgetCtx.Logger().GetLevel()),
+		Timeout:  int64(gadgetCtx.Timeout()),
+	}
+
+	runClient, err := client.RunOCIGadget(connCtx)
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return nil, err
+	}
+
+	controlRequest := &api.GadgetControlRequest{Event: &api.GadgetControlRequest_OciRunRequest{OciRunRequest: runRequest}}
+	err = runClient.Send(controlRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	doneChan := make(chan error)
+
+	var result []byte
+	expectedSeq := uint32(1)
+
+	go func() {
+		dsMap := make(map[uint32]datasource.DataSource)
+		dsNameMap := make(map[string]uint32)
+		initialized := false
+		for {
+			ev, err := runClient.Recv()
+			if err != nil {
+				gadgetCtx.Logger().Debugf("%-20s | runClient returned with %v", target.node, err)
+				if !errors.Is(err, io.EOF) {
+					doneChan <- err
+					return
+				}
+				doneChan <- nil
+				return
+			}
+			switch ev.Type {
+			case api.EventTypeGadgetPayload:
```

Comment:
Can we use `oneof` to avoid handling all the unmarshaling by hand? A new log event message could be created with the log level inside, so the trick of using event to send the log level is not needed anymore. WDYT?

Reply:
I'll look into it again to see if there's a better solution. Right now it's done this way, because we need control over when data is serialized to

* make sure the memory isn't used asynchronously (because it could contain regions that are unsafe to use after the emitter function returns)
* be able to have the sequence counter and drop events if the queue (channel) is full (taking into account the time needed for serialization) - the current design makes it work asynchronously, so serialization has to be done by that time, see point above

Not having the option to drop events could clog memory in all parts of IG and slow everything down.

Reply:
> make sure the memory isn't used asynchronously (because it could contain regions that are unsafe to use after the emitter function returns)

I think we need to make this clear. It means all subscriptions to data source need to process the data on the same callback, otherwise they need to make a copy.